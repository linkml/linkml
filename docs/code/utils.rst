Utils
=====

.. automodule:: linkml.utils.datautils
    :members:
    :undoc-members:

.. automodule:: linkml.utils.validation
    :members:
    :undoc-members:
