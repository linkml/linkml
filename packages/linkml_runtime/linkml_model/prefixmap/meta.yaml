{
   "IAO": "http://purl.obolibrary.org/obo/IAO_",
   "NCIT": "http://ncicb.nci.nih.gov/xml/owl/EVS/Thesaurus.owl#",
   "OIO": "http://www.geneontology.org/formats/oboInOwl#",
   "bibo": "http://purl.org/ontology/bibo/",
   "cdisc": "http://rdf.cdisc.org/mms#",
   "dcterms": "http://purl.org/dc/terms/",
   "linkml": "https://w3id.org/linkml/",
   "oslc": "http://open-services.net/ns/core#",
   "owl": "http://www.w3.org/2002/07/owl#",
   "pav": "http://purl.org/pav/",
   "prov": "http://www.w3.org/ns/prov#",
   "qb": "http://purl.org/linked-data/cube#",
   "qudt": "http://qudt.org/schema/qudt/",
   "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
   "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
   "schema": "http://schema.org/",
   "sh": "http://www.w3.org/ns/shacl#",
   "skos": "http://www.w3.org/2004/02/skos/core#",
   "skosxl": "http://www.w3.org/2008/05/skos-xl#",
   "swrl": "http://www.w3.org/2003/11/swrl#",
   "vann": "https://vocab.org/vann/",
   "xsd": "http://www.w3.org/2001/XMLSchema#",
   "StructuredAlias": {
      "@id": "skosxl:Label"
   },
   "UnitOfMeasure": {
      "@id": "qudt:Unit"
   }
}
