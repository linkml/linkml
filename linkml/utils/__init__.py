from linkml.utils.deprecation import deprecation_warning

__all__ = ["deprecation_warning"]
