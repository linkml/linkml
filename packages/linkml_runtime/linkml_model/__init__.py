from linkml_model.types import String, Integer, Boolean, Float, Double, Decimal, Time, Date, Datetime, Uriorcurie, Uri, \
    Ncname, Objectidentifier, Nodeidentifier
from linkml_model.extensions import Extension, Extensible
from linkml_model.annotations import Annotation, Annotatable
from linkml_model.meta import ElementName, SchemaDefinitionName, TypeDefinitionName, SubsetDefinitionName, DefinitionName, \
    EnumDefinitionName, SlotDefinitionName, ClassDefinitionName, PrefixPrefixPrefix, LocalNameLocalNameSource, \
    AltDescriptionSource, PermissibleValueText, Element, SchemaDefinition, TypeDefinition, SubsetDefinition, \
    Definition, EnumDefinition, SlotDefinition, ClassDefinition, Prefix, LocalName, Example, AltDescription, \
    PermissibleValue, PvFormulaOptions

