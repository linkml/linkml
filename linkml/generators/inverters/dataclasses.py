"""
Create linkml classes from dataclasses.

.. note::
    These linkml classes are strictly "best effort" and do not fully capture
    the full details of a source schema that was used to generate dataclasses with the
    :class:`linkml.generators.PythonGenerator` class.

"""

from collections.abc import Iterable
from dataclasses import MISSING, dataclass, field, fields
from datetime import datetime
from typing import Any, Literal, Type, Union

from linkml_runtime.linkml_model import ClassDefinition, SlotDefinition
from linkml_runtime.linkml_model.meta import AnonymousSlotExpression

python_to_linkml = {
    str: "string",
    int: "integer",
    float: "float",
    bool: "boolean",
    datetime: "datetime",
}


def _is_optional(t: Any) -> bool:
    return getattr(t, "__origin__") is Union and len(t.__args__) == 2 and t.__args__[1] is None


def _strip_optional(t: Any) -> list[Type] | Type:
    if hasattr(t, "__args__"):
        args = [a_type for a_type in t.__args__ if a_type is not None]
        if len(args) == 1:
            return args[0]
        return args
    return t


@dataclass
class DataclassInverter:

    source: Type[dataclass]
    """The dataclass we are inverting"""
    exclude_attributes: Iterable[str] = field(default_factory=list)
    """Attributes that should not be generated for the dataclass"""
    exclude_private: bool = True
    """Exclude private attributes"""

    def ifabsent(self, f: field) -> str | None:
        if f.default is None:
            return None
        elif isinstance(f.default, bool):
            return f"boolean({f.default})"
        elif isinstance(f.default, str):
            return f"string({f.default})"
        elif isinstance(f.default, int):
            return f"integer({f.default})"
        elif isinstance(f.default, float):
            return f"float({f.default})"
        else:
            return None

    def range_or_any_of(
        self, typ: Type
    ) -> dict[Literal["range"], str] | dict[Literal["any_of"], list[AnonymousSlotExpression]]:
        if getattr(typ, "__origin__", None) is Union:
            if _is_optional(typ):
                typ = _strip_optional(typ)
            else:
                # remove optionals
                typ = _strip_optional(typ)
                if len(typ) == 1:
                    typ = typ[0]
                else:
                    any_ofs = [self.range_or_any_of(t) for t in typ]
                    # remove duplicates
                    return {"any_of": [AnonymousSlotExpression(**dict(t)) for t in {tuple(d.items()) for d in any_ofs}]}

        return {"range": python_to_linkml.get(typ, "linkml:Any")}

    def required(self, f: field) -> bool:
        return f.default is MISSING and f.default_factory is MISSING

    def multivalued(self, f: field) -> bool:
        args = _strip_optional(f.type)
        # strip optional returns a literal list, but isinstance(list[type], Iterable) still is True
        return (isinstance(args, Iterable) and not isinstance(args, list)) or (
            isinstance(args, list) and all([isinstance(a, Iterable) for a in args])
        )

    def is_excluded(self, name: str) -> bool:
        """
        True if a slot name is excluded from the generated class
        """
        return name in self.exclude_attributes or (name.startswith("_") and self.exclude_private)

    def render(self) -> ClassDefinition:
        return ClassDefinition(
            name=self.source.__name__,
            description=self.source.__doc__,
            attributes={
                f.name: SlotDefinition(
                    name=f.name,
                    multivalued=self.multivalued(f),
                    ifabsent=self.ifabsent(f),
                    required=self.required(f),
                    **self.range_or_any_of(f.type),
                )
                for f in fields(self.source)
                if not self.is_excluded(f.name)
            },
        )
