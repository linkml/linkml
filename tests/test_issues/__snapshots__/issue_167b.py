# Auto generated from issue_167b.yaml by pythongen.py version: 0.0.1
# Generation date: 2000-01-01T00:00:00
# Schema: annotations_test
#
# id: http://example.org/tests/issue167b
# description:
# license: https://creativecommons.org/publicdomain/zero/1.0/

import dataclasses
import re
from jsonasobj2 import JsonObj, as_dict
from typing import Optional, List, Union, Dict, ClassVar, Any
from dataclasses import dataclass
from linkml_runtime.linkml_model.meta import EnumDefinition, PermissibleValue, PvFormulaOptions

from linkml_runtime.utils.slot import Slot
from linkml_runtime.utils.metamodelcore import empty_list, empty_dict, bnode
from linkml_runtime.utils.yamlutils import YAMLRoot, extended_str, extended_float, extended_int
from linkml_runtime.utils.dataclass_extensions_376 import dataclasses_init_fn_with_kwargs
from linkml_runtime.utils.formatutils import camelcase, underscore, sfx
from linkml_runtime.utils.enumerations import EnumDefinitionImpl
from rdflib import Namespace, URIRef
from linkml_runtime.utils.curienamespace import CurieNamespace


metamodel_version = "1.7.0"
version = None

# Overwrite dataclasses _init_fn to add **kwargs in __init__
dataclasses._init_fn = dataclasses_init_fn_with_kwargs

# Namespaces
EX = CurieNamespace('ex', 'http://example.org/')
LINKML = CurieNamespace('linkml', 'https://w3id.org/linkml/')
DEFAULT_ = EX


# Types

# Class references



class MyClass(YAMLRoot):
    """
    Annotations as tag value pairs. Note that altLabel is defined in the default namespace, not in the SKOS namespace
    """
    _inherited_slots: ClassVar[List[str]] = []

    class_class_uri: ClassVar[URIRef] = EX.MyClass
    class_class_curie: ClassVar[str] = "ex:MyClass"
    class_name: ClassVar[str] = "my class"
    class_model_uri: ClassVar[URIRef] = EX.MyClass


class MyClass2(YAMLRoot):
    """
    -> This form of annotations is a tag/value format, which allows annotations to be annotated. Note, however, that
    the annotation source is NOT a CURIE, rather just a string.
    """
    _inherited_slots: ClassVar[List[str]] = []

    class_class_uri: ClassVar[URIRef] = EX.MyClass2
    class_class_curie: ClassVar[str] = "ex:MyClass2"
    class_name: ClassVar[str] = "my class 2"
    class_model_uri: ClassVar[URIRef] = EX.MyClass2


# Enumerations


# Slots
class slots:
    pass

