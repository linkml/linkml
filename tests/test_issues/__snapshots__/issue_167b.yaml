name: annotations_test
id: http://example.org/tests/issue167b
imports:
- linkml:annotations
- linkml:types
- linkml:extensions
license: https://creativecommons.org/publicdomain/zero/1.0/
prefixes:
  ex:
    prefix_prefix: ex
    prefix_reference: http://example.org/
  linkml:
    prefix_prefix: linkml
    prefix_reference: https://w3id.org/linkml/
emit_prefixes:
- linkml
default_curi_maps:
- semweb_context
default_prefix: ex
default_range: string
types:
  string:
    name: string
    definition_uri: https://w3id.org/linkml/String
    description: A character string
    notes:
    - In RDF serializations, a slot with range of string is treated as a literal or
      type xsd:string.   If you are authoring schemas in LinkML YAML, the type is
      referenced with the lower case "string".
    from_schema: https://w3id.org/linkml/types
    imported_from: linkml:types
    exact_mappings:
    - schema:Text
    base: str
    uri: xsd:string
  integer:
    name: integer
    definition_uri: https://w3id.org/linkml/Integer
    description: An integer
    notes:
    - If you are authoring schemas in LinkML YAML, the type is referenced with the
      lower case "integer".
    from_schema: https://w3id.org/linkml/types
    imported_from: linkml:types
    exact_mappings:
    - schema:Integer
    base: int
    uri: xsd:integer
  boolean:
    name: boolean
    definition_uri: https://w3id.org/linkml/Boolean
    description: A binary (true or false) value
    notes:
    - If you are authoring schemas in LinkML YAML, the type is referenced with the
      lower case "boolean".
    from_schema: https://w3id.org/linkml/types
    imported_from: linkml:types
    exact_mappings:
    - schema:Boolean
    base: Bool
    uri: xsd:boolean
    repr: bool
  float:
    name: float
    definition_uri: https://w3id.org/linkml/Float
    description: A real number that conforms to the xsd:float specification
    notes:
    - If you are authoring schemas in LinkML YAML, the type is referenced with the
      lower case "float".
    from_schema: https://w3id.org/linkml/types
    imported_from: linkml:types
    exact_mappings:
    - schema:Float
    base: float
    uri: xsd:float
  double:
    name: double
    definition_uri: https://w3id.org/linkml/Double
    description: A real number that conforms to the xsd:double specification
    notes:
    - If you are authoring schemas in LinkML YAML, the type is referenced with the
      lower case "double".
    from_schema: https://w3id.org/linkml/types
    imported_from: linkml:types
    close_mappings:
    - schema:Float
    base: float
    uri: xsd:double
  decimal:
    name: decimal
    definition_uri: https://w3id.org/linkml/Decimal
    description: A real number with arbitrary precision that conforms to the xsd:decimal
      specification
    notes:
    - If you are authoring schemas in LinkML YAML, the type is referenced with the
      lower case "decimal".
    from_schema: https://w3id.org/linkml/types
    imported_from: linkml:types
    broad_mappings:
    - schema:Number
    base: Decimal
    uri: xsd:decimal
  time:
    name: time
    definition_uri: https://w3id.org/linkml/Time
    description: A time object represents a (local) time of day, independent of any
      particular day
    notes:
    - URI is dateTime because OWL reasoners do not work with straight date or time
    - If you are authoring schemas in LinkML YAML, the type is referenced with the
      lower case "time".
    from_schema: https://w3id.org/linkml/types
    imported_from: linkml:types
    exact_mappings:
    - schema:Time
    base: XSDTime
    uri: xsd:time
    repr: str
  date:
    name: date
    definition_uri: https://w3id.org/linkml/Date
    description: a date (year, month and day) in an idealized calendar
    notes:
    - URI is dateTime because OWL reasoners don't work with straight date or time
    - If you are authoring schemas in LinkML YAML, the type is referenced with the
      lower case "date".
    from_schema: https://w3id.org/linkml/types
    imported_from: linkml:types
    exact_mappings:
    - schema:Date
    base: XSDDate
    uri: xsd:date
    repr: str
  datetime:
    name: datetime
    definition_uri: https://w3id.org/linkml/Datetime
    description: The combination of a date and time
    notes:
    - If you are authoring schemas in LinkML YAML, the type is referenced with the
      lower case "datetime".
    from_schema: https://w3id.org/linkml/types
    imported_from: linkml:types
    exact_mappings:
    - schema:DateTime
    base: XSDDateTime
    uri: xsd:dateTime
    repr: str
  date_or_datetime:
    name: date_or_datetime
    definition_uri: https://w3id.org/linkml/DateOrDatetime
    description: Either a date or a datetime
    notes:
    - If you are authoring schemas in LinkML YAML, the type is referenced with the
      lower case "date_or_datetime".
    from_schema: https://w3id.org/linkml/types
    imported_from: linkml:types
    base: str
    uri: linkml:DateOrDatetime
    repr: str
  uriorcurie:
    name: uriorcurie
    definition_uri: https://w3id.org/linkml/Uriorcurie
    description: a URI or a CURIE
    notes:
    - If you are authoring schemas in LinkML YAML, the type is referenced with the
      lower case "uriorcurie".
    from_schema: https://w3id.org/linkml/types
    imported_from: linkml:types
    base: URIorCURIE
    uri: xsd:anyURI
    repr: str
  curie:
    name: curie
    definition_uri: https://w3id.org/linkml/Curie
    conforms_to: https://www.w3.org/TR/curie/
    description: a compact URI
    notes:
    - If you are authoring schemas in LinkML YAML, the type is referenced with the
      lower case "curie".
    comments:
    - in RDF serializations this MUST be expanded to a URI
    - in non-RDF serializations MAY be serialized as the compact representation
    from_schema: https://w3id.org/linkml/types
    imported_from: linkml:types
    base: Curie
    uri: xsd:string
    repr: str
  uri:
    name: uri
    definition_uri: https://w3id.org/linkml/Uri
    conforms_to: https://www.ietf.org/rfc/rfc3987.txt
    description: a complete URI
    notes:
    - If you are authoring schemas in LinkML YAML, the type is referenced with the
      lower case "uri".
    comments:
    - in RDF serializations a slot with range of uri is treated as a literal or type
      xsd:anyURI unless it is an identifier or a reference to an identifier, in which
      case it is translated directly to a node
    from_schema: https://w3id.org/linkml/types
    imported_from: linkml:types
    close_mappings:
    - schema:URL
    base: URI
    uri: xsd:anyURI
    repr: str
  ncname:
    name: ncname
    definition_uri: https://w3id.org/linkml/Ncname
    description: Prefix part of CURIE
    notes:
    - If you are authoring schemas in LinkML YAML, the type is referenced with the
      lower case "ncname".
    from_schema: https://w3id.org/linkml/types
    imported_from: linkml:types
    base: NCName
    uri: xsd:string
    repr: str
  objectidentifier:
    name: objectidentifier
    definition_uri: https://w3id.org/linkml/Objectidentifier
    description: A URI or CURIE that represents an object in the model.
    notes:
    - If you are authoring schemas in LinkML YAML, the type is referenced with the
      lower case "objectidentifier".
    comments:
    - Used for inheritance and type checking
    from_schema: https://w3id.org/linkml/types
    imported_from: linkml:types
    base: ElementIdentifier
    uri: shex:iri
    repr: str
  nodeidentifier:
    name: nodeidentifier
    definition_uri: https://w3id.org/linkml/Nodeidentifier
    description: A URI, CURIE or BNODE that represents a node in a model.
    notes:
    - If you are authoring schemas in LinkML YAML, the type is referenced with the
      lower case "nodeidentifier".
    from_schema: https://w3id.org/linkml/types
    imported_from: linkml:types
    base: NodeIdentifier
    uri: shex:nonLiteral
    repr: str
  jsonpointer:
    name: jsonpointer
    definition_uri: https://w3id.org/linkml/Jsonpointer
    conforms_to: https://datatracker.ietf.org/doc/html/rfc6901
    description: A string encoding a JSON Pointer. The value of the string MUST conform
      to JSON Point syntax and SHOULD dereference to a valid object within the current
      instance document when encoded in tree form.
    notes:
    - If you are authoring schemas in LinkML YAML, the type is referenced with the
      lower case "jsonpointer".
    from_schema: https://w3id.org/linkml/types
    imported_from: linkml:types
    base: str
    uri: xsd:string
    repr: str
  jsonpath:
    name: jsonpath
    definition_uri: https://w3id.org/linkml/Jsonpath
    conforms_to: https://www.ietf.org/archive/id/draft-goessner-dispatch-jsonpath-00.html
    description: A string encoding a JSON Path. The value of the string MUST conform
      to JSON Point syntax and SHOULD dereference to zero or more valid objects within
      the current instance document when encoded in tree form.
    notes:
    - If you are authoring schemas in LinkML YAML, the type is referenced with the
      lower case "jsonpath".
    from_schema: https://w3id.org/linkml/types
    imported_from: linkml:types
    base: str
    uri: xsd:string
    repr: str
  sparqlpath:
    name: sparqlpath
    definition_uri: https://w3id.org/linkml/Sparqlpath
    conforms_to: https://www.w3.org/TR/sparql11-query/#propertypaths
    description: A string encoding a SPARQL Property Path. The value of the string
      MUST conform to SPARQL syntax and SHOULD dereference to zero or more valid objects
      within the current instance document when encoded as RDF.
    notes:
    - If you are authoring schemas in LinkML YAML, the type is referenced with the
      lower case "sparqlpath".
    from_schema: https://w3id.org/linkml/types
    imported_from: linkml:types
    base: str
    uri: xsd:string
    repr: str
slots:
  annotations:
    name: annotations
    definition_uri: https://w3id.org/linkml/annotations
    description: a collection of tag/text tuples with the semantics of OWL Annotation
    from_schema: https://w3id.org/linkml/annotations
    imported_from: linkml:annotations
    is_a: extensions
    domain: annotatable
    slot_uri: linkml:annotations
    multivalued: true
    owner: annotation
    domain_of:
    - annotatable
    - annotation
    range: annotation
    inlined: true
  extensions:
    name: extensions
    definition_uri: https://w3id.org/linkml/extensions
    description: a tag/text tuple attached to an arbitrary element
    from_schema: https://w3id.org/linkml/extensions
    imported_from: linkml:extensions
    domain: extensible
    slot_uri: linkml:extensions
    multivalued: true
    owner: extensible
    domain_of:
    - extension
    - extensible
    range: extension
    inlined: true
  extension_tag:
    name: extension_tag
    definition_uri: https://w3id.org/linkml/extension_tag
    description: a tag associated with an extension
    from_schema: https://w3id.org/linkml/extensions
    imported_from: linkml:extensions
    domain: extension
    slot_uri: linkml:tag
    key: true
    alias: tag
    owner: extension
    domain_of:
    - extension
    range: uriorcurie
    required: true
  extension_value:
    name: extension_value
    definition_uri: https://w3id.org/linkml/extension_value
    description: the actual annotation
    from_schema: https://w3id.org/linkml/extensions
    imported_from: linkml:extensions
    domain: extension
    slot_uri: linkml:value
    alias: value
    owner: extension
    domain_of:
    - extension
    range: AnyValue
    required: true
    inlined: true
classes:
  my class:
    name: my class
    definition_uri: http://example.org/MyClass
    annotations:
      SKOS:prefLabel:
        tag: SKOS:prefLabel
        value: a label
      altLabel:
        tag: altLabel
        value: some other label
    description: Annotations as tag value pairs. Note that altLabel is defined in
      the default namespace, not in the SKOS namespace
    from_schema: http://example.org/tests/issue167b
    class_uri: ex:MyClass
  my class 2:
    name: my class 2
    definition_uri: http://example.org/MyClass2
    annotations:
      prefLabel:
        tag: prefLabel
        value: a preferred label
        annotations:
          source:
            tag: source
            value: PMID:12345
    description: -> This form of annotations is a tag/value format, which allows annotations
      to be annotated.  Note, however, that the annotation source is NOT a CURIE,
      rather just a string.
    from_schema: http://example.org/tests/issue167b
    class_uri: ex:MyClass2
  annotatable:
    name: annotatable
    definition_uri: https://w3id.org/linkml/Annotatable
    description: mixin for classes that support annotations
    from_schema: https://w3id.org/linkml/annotations
    imported_from: linkml:annotations
    mixin: true
    slots:
    - annotations
    class_uri: linkml:Annotatable
  annotation:
    name: annotation
    definition_uri: https://w3id.org/linkml/Annotation
    description: a tag/value pair with the semantics of OWL Annotation
    from_schema: https://w3id.org/linkml/annotations
    imported_from: linkml:annotations
    is_a: extension
    mixins:
    - annotatable
    slots:
    - extension_tag
    - extension_value
    - extensions
    - annotations
    class_uri: linkml:Annotation
  AnyValue:
    name: AnyValue
    definition_uri: https://w3id.org/linkml/AnyValue
    from_schema: https://w3id.org/linkml/extensions
    imported_from: linkml:extensions
    mappings:
    - linkml:Any
    class_uri: linkml:Any
  extension:
    name: extension
    definition_uri: https://w3id.org/linkml/Extension
    description: a tag/value pair used to add non-model information to an entry
    from_schema: https://w3id.org/linkml/extensions
    imported_from: linkml:extensions
    slots:
    - extension_tag
    - extension_value
    - extensions
    class_uri: linkml:Extension
  extensible:
    name: extensible
    definition_uri: https://w3id.org/linkml/Extensible
    description: mixin for classes that support extension
    from_schema: https://w3id.org/linkml/extensions
    imported_from: linkml:extensions
    mixin: true
    slots:
    - extensions
    class_uri: linkml:Extensible
metamodel_version: 1.7.0
source_file: issue_167b.yaml
source_file_date: '2000-01-01T00:00:00'
source_file_size: 1
generation_date: '2000-01-01T00:00:00'
